src/platform/packages/shared/kbn-esql-validation-autocomplete/src/autocomplete/helper.ts

```diff
@@ -1038,3 +1044,118 @@ export function isExpressionComplete(
     !(isNullMatcher.test(innerText) || isNotNullMatcher.test(innerText))
   );
 }
+
+export function getSourceSuggestions(sources: ESQLSourceResult[]) {
+  // hide indexes that start with .
+  return buildSourcesDefinitions(
+    sources
+      .filter(({ hidden }) => !hidden)
+      .map(({ name, dataStreams, title, type }) => {
+        return { name, isIntegration: Boolean(dataStreams && dataStreams.length), title, type };
+      })
+  );
+}
+
+export async function additionalSourcesSuggestions(
+  queryText: string,
+  sources: ESQLSourceResult[],
+  recommendedQuerySuggestions: SuggestionRawDefinition[]
+) {
+  const canRemoveQuote = queryText.includes('"');
+  const suggestionsToAdd = await handleFragment(
+    queryText,
+    (fragment) =>
+      sourceExists(fragment, new Set(sources.map(({ name: sourceName }) => sourceName))),
+    (_fragment, rangeToReplace) => {
+      return getSourceSuggestions(sources).map((suggestion) => ({
+        ...suggestion,
+        rangeToReplace,
+      }));
+    },
+    (fragment, rangeToReplace) => {
+      const exactMatch = sources.find(({ name: _name }) => _name === fragment);
+      if (exactMatch?.dataStreams) {
+        // this is an integration name, suggest the datastreams
+        const definitions = buildSourcesDefinitions(
+          exactMatch.dataStreams.map(({ name }) => ({ name, isIntegration: false }))
+        );
+
+        return canRemoveQuote ? removeQuoteForSuggestedSources(definitions) : definitions;
+      } else {
+        const _suggestions: SuggestionRawDefinition[] = [
+          {
+            ...pipeCompleteItem,
+            filterText: fragment,
+            text: fragment + ' | ',
+            command: TRIGGER_SUGGESTION_COMMAND,
+            rangeToReplace,
+          },
+          {
+            ...commaCompleteItem,
+            filterText: fragment,
+            text: fragment + ', ',
+            command: TRIGGER_SUGGESTION_COMMAND,
+            rangeToReplace,
+          },
+          {
+            ...metadataSuggestion,
+            filterText: fragment,
+            text: fragment + ' METADATA ',
+            rangeToReplace,
+          },
+          ...recommendedQuerySuggestions.map((suggestion) => ({
+            ...suggestion,
+            rangeToReplace,
+            filterText: fragment,
+            text: fragment + suggestion.text,
+          })),
+        ];
+        return _suggestions;
+      }
+    }
+  );
+  return suggestionsToAdd;
+}
+
+export const specialIndicesToSuggestions = (
```

Comment:
This could use a comment. What are "special" indices... why do they get special logic?

Reply:
Sure done https://github.com/elastic/kibana/pull/219901/commits/f4bbffdd9fc26f9a15d944280161f0a4c5324800